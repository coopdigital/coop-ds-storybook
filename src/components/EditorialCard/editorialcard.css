@media (min-width: 48rem) and (max-width: 63.99rem) {
  .coop-c-editorialcard--horizontalOnTablet {
    display: flex;
  }

  .coop-c-editorialcard--horizontalOnTablet .coop-c-editorialcard__inner {
    display: flex;
    flex-direction: row;
  }

  .coop-c-editorialcard--horizontalOnTablet .coop-c-editorialcard__media {
    overflow: hidden;
    flex-basis: 31.6506309148%;
    width: 31.6506309148%;
  }

  .coop-c-editorialcard--horizontalOnTablet .coop-c-editorialcard__image,
  .coop-c-editorialcard--horizontalOnTablet .coop-c-editorialcard__image img {
    height: 100%;
    width: 100%;
  }

  .coop-c-editorialcard--horizontalOnTablet .coop-c-editorialcard__image img {
    object-fit: cover;
  }
}
