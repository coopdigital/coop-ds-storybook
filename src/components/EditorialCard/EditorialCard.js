import React from "react";
import PropTypes from "prop-types";
// import sanitizeHtml from "../../utils/sanitizeHtml";
import classNames from "../../utils/classNames";

// import Img from "../components/Img";
import "@coopdigital/shared-component--editorialcard/dist/editorialCard.css";
import "./editorialcard.css";

const EditorialCard = ({
  title,
  subtitle,
  content,
  link,
  mediaComponent,
  horizontal,
  horizontalOnTablet,
  className,
}) => {
  const classes = classNames("coop-c-editorialcard", [
    horizontal && `coop-c-editorialcard--horizontal`,
    horizontalOnTablet && `coop-c-editorialcard--horizontalOnTablet`,
    className,
  ]);
  return (
    <article className={classes}>
      <a
        href={link}
        className="coop-c-editorialcard__link"
        data-contenttype="Card|Editorial"
        data-contentparent=""
        data-linktext=""
      >
        <div className="coop-c-editorialcard__inner">
          {mediaComponent && (
            <div className="coop-c-editorialcard__media">
              {mediaComponent}
            </div>
          )}

          <div className="coop-c-editorialcard__content">
            <header className="coop-c-editorialcard__header">
              {subtitle && (
                <p className="coop-c-editorialcard__subtitle">{subtitle}</p>
              )}
              {title && (
                <h3 className="coop-c-editorialcard__title">
                  <span>{title}</span>
                </h3>
              )}
            </header>
            {content && (
              <div className="coop-c-editorialcard__body" dangerouslySetInnerHTML={{ __html: content }}/>
            )}
          </div>
        </div>
      </a>
    </article>
  );
};

EditorialCard.defaultProps = {
  subtitle: null,
};

EditorialCard.propTypes = {
  title: PropTypes.string.isRequired,
  subtitle: PropTypes.string,
  content: PropTypes.string.isRequired,
  link: PropTypes.string.isRequired,
  /** mediaComponent - is optional and usually take an image in form of a ResponsiveImage react component */
  mediaComponent: PropTypes.node,
};

export default EditorialCard;
