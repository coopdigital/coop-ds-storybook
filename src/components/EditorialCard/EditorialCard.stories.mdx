import {
  ArgsTable,
  Meta,
  Story,
  Canvas,
  Source,
  Preview,
} from "@storybook/addon-docs/blocks";
import EditorialCard from "./EditorialCard";
import ResponsiveImage from "../ResponsiveImage";

import Grid from "../Grid";
import GridItem from "../GridItem";

<!-- Set up mediacomponent  -->

export const MediaComponent = () => {
  const src =
    "https://images.ctfassets.net/bffxiku554r1/7XDwxNP8ioupeMdJZxNS8/d23d2b802d6be0ecbcb7abd1ff33b159/coop-supporting-local-communities.png?fm=png&q=60&w=600&h=338";
  const alt = "How Co-op is supporting local communities";
  const figureClass = 'coop-c-editorialcard__media';
  const pictureClass = 'coop-c-editorialcard__image';
  return <ResponsiveImage src={src} alt={alt} figureClass={figureClass} pictureClass={pictureClass} />;
};

export const testContent = `<p>hello and welcome to the content of a card <span>I'm in a span.</span></p>
<p>another paragraph tag</p>`;

<Meta
  title="Component/EditorialCard"
  component={EditorialCard}
  args={{
    title: "Editorial card title",
    content: `<p>hello and welcome to the content of a card. <span>I'm in a span.</span></p>
<p>Another paragraph tag</p>`,
    link:
      "https://coop-design-system.herokuapp.com/pattern-library/components/cards/editorial-card.html",
  }}
/>

# EditorialCard
NOTE:: Using the ResponsiveImage component as a subcomponent does work here yet!
An editorial card contains a small piece of prose information on one subject that people can get more information about by tapping or clicking on the card

## Default EditorialCard Example
A test canvas to show the component and it's underlying code.
A canvas can include stories or just plain components like the one below.

<Canvas>
  <EditorialCard
    title="card title"
    subtitle="card subtitle"
    content={testContent}
  />
</Canvas>

<Story name="Basic">{(args) => <EditorialCard {...args} />}</Story>

<Story name="With subtitle" args={{ subtitle: "A card subtitle" }}>
  {(args) => <EditorialCard {...args} />}
</Story>

<Story name="With image" args={{ mediaComponent: <MediaComponent /> }}>
  {(args) => <EditorialCard {...args} />}
</Story>

<Story
  name="With image and subtitle"
  args={{
    subtitle: "A card subtitle",
    mediaComponent: <MediaComponent />,
  }}
>
  {(args) => <EditorialCard {...args} />}
</Story>


<Story
  name="With image horizontal"
  args={{
    horizontal: true,
    mediaComponent: <MediaComponent />,
  }}
>
  {(args) => <EditorialCard {...args} />}
</Story>

<Story
  name="In 2 columns vertical"
  args={{
    mediaComponent: <MediaComponent />,
  }}
>{(args) =>
  <Grid>
    <GridItem md="6" key="1">
      <EditorialCard {...args} />
    </GridItem>
    <GridItem md="6" key="2">
      <EditorialCard {...args} />
    </GridItem>
  </Grid>
  }
</Story>


<Story
  name="In 3 columns vertical"
  args={{
    mediaComponent: <MediaComponent />,
  }}
>{(args) =>
  <Grid>
    <GridItem md="4" key="1">
      <EditorialCard {...args} />
    </GridItem>
    <GridItem md="4" key="2">
      <EditorialCard {...args} />
    </GridItem>
    <GridItem md="4" key="3">
      <EditorialCard {...args} />
    </GridItem>
  </Grid>
  }
</Story>

<Story
  name="In 4 columns vertical"
  args={{
    mediaComponent: <MediaComponent />,
  }}
>{(args) =>
  <Grid>
    <GridItem md="3" key="1">
      <EditorialCard {...args} />
    </GridItem>
    <GridItem md="3" key="2">
      <EditorialCard {...args} />
    </GridItem>
    <GridItem md="3" key="3">
      <EditorialCard {...args} />
    </GridItem>
    <GridItem md="3" key="4">
      <EditorialCard {...args} />
    </GridItem>
  </Grid>
  }
</Story>


<Story
  name="In 2 columns horizontal"
  args={{
    horizontal: true,
    mediaComponent: <MediaComponent />,
  }}
>{(args) =>
  <Grid>
    <GridItem sm="6" md="12" lg="6" key="1">
      <EditorialCard {...args} />
    </GridItem>
    <GridItem sm="6" md="12" lg="6" key="2">
      <EditorialCard {...args} />
    </GridItem>
  </Grid>
  }
</Story>


<Story
  name="In 2 columns horizontalOnTablet"
  args={{
    horizontalOnTablet: true,
    mediaComponent: <MediaComponent />,
  }}
>{(args) =>
  <Grid>
    <GridItem sm="6" md="12" lg="6"  key="1">
      <EditorialCard {...args} />
    </GridItem>
    <GridItem sm="6" md="12" lg="6"  key="2">
      <EditorialCard {...args} />
    </GridItem>
  </Grid>
  }
</Story>


<Story
  name="In 3 columns horizontalOnTablet"
  args={{
    horizontalOnTablet: true,
    mediaComponent: <MediaComponent />,
  }}
>{(args) =>
  <Grid>
    <GridItem md="12" lg="4"  key="1">
      <EditorialCard {...args} />
    </GridItem>
    <GridItem md="12" lg="4" key="2">
      <EditorialCard {...args} />
    </GridItem>
    <GridItem md="12" lg="4" key="3">
      <EditorialCard {...args} />
    </GridItem>
  </Grid>
  }
</Story>