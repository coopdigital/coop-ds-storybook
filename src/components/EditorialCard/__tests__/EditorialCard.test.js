import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { graphql } from "gatsby";

import EditorialCard from "../EditorialCard";

//mock data

describe("Editorial Card test", () => {
  it("renders correctly", () => {
    const component = render(
      <EditorialCard
        title="Editorialcard title"
        link="#"
        content="content test"
      />
    );
    expect(component).toMatchSnapshot();
  });

  // it should render as horizontal css with horizontal prop added
  it("should render as horizontal css with horizontal prop added", () => {
    const { container } = render(
      <EditorialCard
        horizontal
        title="Editorial card horizontal"
        link="#"
        content="content test"
      />
    );
    expect(container.firstChild).toHaveClass(
      "coop-c-editorialcard--horizontal"
    );
    // expect(container).toMatchSnapshot();
  });

  // it should allow further className extenstions
  it("should allow further className extenstions", () => {
    const { container } = render(
      <EditorialCard
        title="Editorial card"
        link="#"
        content="content test"
        className="test-classname test-classname--1"
      />
    );
    expect(container.firstChild).toHaveClass("test-classname");
    expect(container.firstChild).toHaveClass("test-classname--1");
  });

  //it should add relevant class when horizontal-at-md prop is added
  it("should add relevant class when horizontal-at-md prop is added", () => {
    const { container } = render(
      <EditorialCard
        horizontalOnTablet
        title="Editorial card horizontal"
        link="#"
        content="content test"
      />
    );

    expect(container.firstChild).toHaveClass(
      "coop-c-editorialcard--horizontalOnTablet"
    );
  });

  // test gatsby-image <Img/> component??

  // it should render with image when prop has value
  // it should render without error when no image is supplied

  // test gatsby <Link/>
  // it should error when no link is supplied
});
