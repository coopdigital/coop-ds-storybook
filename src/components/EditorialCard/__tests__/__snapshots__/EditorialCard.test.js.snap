// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editorial Card test renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <article
        class="coop-c-editorialcard"
      >
        <a
          class="coop-c-editorialcard__link"
          data-contentparent=""
          data-contenttype="Card|Editorial"
          data-linktext=""
          href="#"
        >
          <div
            class="coop-c-editorialcard__inner"
          >
            <div
              class="coop-c-editorialcard__content"
            >
              <header
                class="coop-c-editorialcard__header"
              >
                <h3
                  class="coop-c-editorialcard__title"
                >
                  <span>
                    Editorialcard title
                  </span>
                </h3>
              </header>
              <div
                class="coop-c-editorialcard__body"
              >
                <p>
                  content test
                </p>
              </div>
            </div>
          </div>
        </a>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="coop-c-editorialcard"
    >
      <a
        class="coop-c-editorialcard__link"
        data-contentparent=""
        data-contenttype="Card|Editorial"
        data-linktext=""
        href="#"
      >
        <div
          class="coop-c-editorialcard__inner"
        >
          <div
            class="coop-c-editorialcard__content"
          >
            <header
              class="coop-c-editorialcard__header"
            >
              <h3
                class="coop-c-editorialcard__title"
              >
                <span>
                  Editorialcard title
                </span>
              </h3>
            </header>
            <div
              class="coop-c-editorialcard__body"
            >
              <p>
                content test
              </p>
            </div>
          </div>
        </div>
      </a>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
